In [6]:
import pandas as pd

### Wind Energy Generation Data

In [78]:
df= pd.read_csv('/Users/tovi/Library/Mobile Documents/com~apple~CloudDocs/texas-renewables-analysis/clean_data/ercot_wind.csv')
df.head()

timestamp  wind_system  wind_coast  wind_south  wind_west  \
0  2021-01-01 01:00:00     14012.32     1975.35     1246.65    7623.01   
1  2021-01-01 02:00:00     14315.26     2023.05      980.89    7778.31   
2  2021-01-01 03:00:00     14247.88     1819.02      742.39    8068.91   
3  2021-01-01 04:00:00     13273.42     1426.17      750.18    7543.59   
4  2021-01-01 05:00:00     12248.01     1280.43      825.16    6667.16   

   wind_north  
0      146.97  
1      174.87  
2      211.95  
3      186.87  
4      169.53

In [80]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25463 entries, 0 to 25462
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   timestamp    25463 non-null  object 
 1   wind_system  25456 non-null  float64
 2   wind_coast   25456 non-null  float64
 3   wind_south   25456 non-null  float64
 4   wind_west    25456 non-null  float64
 5   wind_north   25456 non-null  float64
dtypes: float64(5), object(1)
memory usage: 1.2+ MB


In [ ]:
"""
CREATE TABLE wind_data (timestamp TIMESTAMP NOT NULL,
						wind_system DECIMAL,
						wind_coast DECIMAL,
						wind_south DECIMAL,
						wind_west DECIMAL,
						wind_north DECIMAL
						);
"""

Returns 25463 rows as expected (Null values within the timestamp column were not allowed.

### Temperature Data

In [47]:
wind_temp_df = pd.read_csv('/Users/tovi/Downloads/era5_by_region.csv')
wind_temp_df.head()

time  number  temp_2m_south  wind_u_100m_south  \
0  2021-01-01 00:00:00       0      280.44113           8.874052   
1  2021-01-01 01:00:00       0      280.06485           8.833785   
2  2021-01-01 02:00:00       0      279.50992           8.720156   
3  2021-01-01 03:00:00       0      278.65192           8.357309   
4  2021-01-01 04:00:00       0      278.41907           8.048112   

   wind_v_100m_south  number.1  temp_2m_north  wind_u_100m_north  \
0          -4.125525         0      275.47700          -0.758140   
1          -4.142875         0      275.49615          -0.381892   
2          -3.938205         0      275.44696           0.146601   
3          -3.723078         0      275.24594           0.887693   
4          -3.574828         0      275.12744           1.753668   

   wind_v_100m_north  number.2  temp_2m_east  wind_u_100m_east  \
0          -8.064275         0     285.33070          8.095841   
1          -7.454720         0     284.06274          9.073054   
2          -6.802784         0     282.77908          8.665540   
3          -6.319172         0     281.25110          7.918151   
4          -5.496035         0     280.96722          8.195437   

   wind_v_100m_east  number.3  temp_2m_west  wind_u_100m_west  \
0          5.828926         0     274.88324          3.239945   
1          5.087430         0     274.58267          3.051376   
2          4.598661         0     273.86377          2.987516   
3          4.829106         0     273.21588          2.977859   
4          5.092289         0     272.90448          3.114550   

   wind_v_100m_west  
0         -4.025888  
1         -3.864054  
2         -3.852034  
3         -3.923165  
4         -3.790921

In [49]:
columns_to_drop = ['number', 'number.1', 'number.2', 'number.3']
wind_df_dropped = wind_temp_df.drop(columns=columns_to_drop, axis=1)

In [51]:
wind_df_dropped.head()

time  temp_2m_south  wind_u_100m_south  wind_v_100m_south  \
0  2021-01-01 00:00:00      280.44113           8.874052          -4.125525   
1  2021-01-01 01:00:00      280.06485           8.833785          -4.142875   
2  2021-01-01 02:00:00      279.50992           8.720156          -3.938205   
3  2021-01-01 03:00:00      278.65192           8.357309          -3.723078   
4  2021-01-01 04:00:00      278.41907           8.048112          -3.574828   

   temp_2m_north  wind_u_100m_north  wind_v_100m_north  temp_2m_east  \
0      275.47700          -0.758140          -8.064275     285.33070   
1      275.49615          -0.381892          -7.454720     284.06274   
2      275.44696           0.146601          -6.802784     282.77908   
3      275.24594           0.887693          -6.319172     281.25110   
4      275.12744           1.753668          -5.496035     280.96722   

   wind_u_100m_east  wind_v_100m_east  temp_2m_west  wind_u_100m_west  \
0          8.095841          5.828926     274.88324          3.239945   
1          9.073054          5.087430     274.58267          3.051376   
2          8.665540          4.598661     273.86377          2.987516   
3          7.918151          4.829106     273.21588          2.977859   
4          8.195437          5.092289     272.90448          3.114550   

   wind_v_100m_west  
0         -4.025888  
1         -3.864054  
2         -3.852034  
3         -3.923165  
4         -3.790921

In [53]:
# List of temperature columns to convert
temp_columns = ['temp_2m_south', 'temp_2m_north', 'temp_2m_east', 'temp_2m_west']

# Convert from kelvin to celsius
wind_df_dropped[temp_columns] = wind_df_dropped[temp_columns] - 273.15

wind_df_dropped.head()

time  temp_2m_south  wind_u_100m_south  wind_v_100m_south  \
0  2021-01-01 00:00:00        7.29113           8.874052          -4.125525   
1  2021-01-01 01:00:00        6.91485           8.833785          -4.142875   
2  2021-01-01 02:00:00        6.35992           8.720156          -3.938205   
3  2021-01-01 03:00:00        5.50192           8.357309          -3.723078   
4  2021-01-01 04:00:00        5.26907           8.048112          -3.574828   

   temp_2m_north  wind_u_100m_north  wind_v_100m_north  temp_2m_east  \
0        2.32700          -0.758140          -8.064275      12.18070   
1        2.34615          -0.381892          -7.454720      10.91274   
2        2.29696           0.146601          -6.802784       9.62908   
3        2.09594           0.887693          -6.319172       8.10110   
4        1.97744           1.753668          -5.496035       7.81722   

   wind_u_100m_east  wind_v_100m_east  temp_2m_west  wind_u_100m_west  \
0          8.095841          5.828926       1.73324          3.239945   
1          9.073054          5.087430       1.43267          3.051376   
2          8.665540          4.598661       0.71377          2.987516   
3          7.918151          4.829106       0.06588          2.977859   
4          8.195437          5.092289      -0.24552          3.114550   

   wind_v_100m_west  
0         -4.025888  
1         -3.864054  
2         -3.852034  
3         -3.923165  
4         -3.790921

In [55]:
# converting from celcius to fahrenheit
wind_df_dropped[temp_columns] = (wind_df_dropped[temp_columns] * 9/5) + 32
wind_df_dropped.head()

time  temp_2m_south  wind_u_100m_south  wind_v_100m_south  \
0  2021-01-01 00:00:00      45.124034           8.874052          -4.125525   
1  2021-01-01 01:00:00      44.446730           8.833785          -4.142875   
2  2021-01-01 02:00:00      43.447856           8.720156          -3.938205   
3  2021-01-01 03:00:00      41.903456           8.357309          -3.723078   
4  2021-01-01 04:00:00      41.484326           8.048112          -3.574828   

   temp_2m_north  wind_u_100m_north  wind_v_100m_north  temp_2m_east  \
0      36.188600          -0.758140          -8.064275     53.925260   
1      36.223070          -0.381892          -7.454720     51.642932   
2      36.134528           0.146601          -6.802784     49.332344   
3      35.772692           0.887693          -6.319172     46.581980   
4      35.559392           1.753668          -5.496035     46.070996   

   wind_u_100m_east  wind_v_100m_east  temp_2m_west  wind_u_100m_west  \
0          8.095841          5.828926     35.119832          3.239945   
1          9.073054          5.087430     34.578806          3.051376   
2          8.665540          4.598661     33.284786          2.987516   
3          7.918151          4.829106     32.118584          2.977859   
4          8.195437          5.092289     31.558064          3.114550   

   wind_v_100m_west  
0         -4.025888  
1         -3.864054  
2         -3.852034  
3         -3.923165  
4         -3.790921

In [63]:
wind_df_dropped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26280 entries, 0 to 26279
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   time               26280 non-null  object 
 1   temp_2m_south      26280 non-null  float64
 2   wind_u_100m_south  26280 non-null  float64
 3   wind_v_100m_south  26280 non-null  float64
 4   temp_2m_north      26280 non-null  float64
 5   wind_u_100m_north  26280 non-null  float64
 6   wind_v_100m_north  26280 non-null  float64
 7   temp_2m_east       26280 non-null  float64
 8   wind_u_100m_east   26280 non-null  float64
 9   wind_v_100m_east   26280 non-null  float64
 10  temp_2m_west       26280 non-null  float64
 11  wind_u_100m_west   26280 non-null  float64
 12  wind_v_100m_west   26280 non-null  float64
dtypes: float64(12), object(1)
memory usage: 2.6+ MB


In [59]:
# saving csv to be joined with energy generation and hourly load data files using SQL
wind_df_dropped.to_csv('/Users/tovi/Downloads/wind_temps_converted.csv', index=False)

In [ ]:
"""
CREATE TABLE wind_temp_data (time TIMESTAMP NOT NULL,
							  temp_2m_south DECIMAL(5,2),
							  wind_u_south DECIMAL(14,10),
							  wind_v_south DECIMAL(14,10),
							  temp_2m_north DECIMAL (5,2),
							  wind_u_north DECIMAL(14,10),
							  wind_v_north DECIMAL(14,10),
							  temp_2m_east DECIMAL (5,2),
							  wind_u_east DECIMAL(14,10),
							  wind_v_east DECIMAL(14,10),
							  temp_2m_west DECIMAL (5,2),
							  wind_u_west DECIMAL(14,10),
							  wind_v_west DECIMAL(14,10)
							  );
"""

Returns 26280 rows as expected. DECIMAL data type precision and scale set to (14,10) to contain large windspeed float values. For temperature values it was set to (5,2) so that all values are rounded to 2 decimal places.

### Hourly Load Data

In [85]:
load_df = pd.read_csv('/Users/tovi/Library/Mobile Documents/com~apple~CloudDocs/texas-renewables-analysis/clean_data/ercot_hourly_load.csv')
load_df.head()

timestamp      coast      east   farwest   north northcentral  \
0  2021-01-01 00:00:00  11,047.02  1,517.44  3,978.22  838.35    13,575.28   
1  2021-01-01 01:00:00  10,954.47  1,523.68  3,975.76  825.12    13,340.34   
2  2021-01-01 02:00:00  10,883.66  1,519.80  3,969.81  819.55    13,311.63   
3  2021-01-01 03:00:00  10,926.48  1,523.47  3,983.49  820.61    13,356.44   
4  2021-01-01 04:00:00  11,019.80  1,565.63  3,980.94  831.61    13,528.65   

      south southcentral      west     system  
0  3,545.09     7,778.26  1,440.19  43,719.85  
1  3,586.55     7,693.74  1,421.39  43,321.05  
2  3,586.12     7,566.95  1,405.54  43,063.07  
3  3,570.59     7,506.69  1,402.28  43,090.06  
4  3,590.90     7,563.97  1,405.09  43,486.59

In [87]:
load_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26280 entries, 0 to 26279
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   timestamp     26280 non-null  object
 1   coast         26280 non-null  object
 2   east          26280 non-null  object
 3   farwest       26280 non-null  object
 4   north         26280 non-null  object
 5   northcentral  26280 non-null  object
 6   south         26280 non-null  object
 7   southcentral  26280 non-null  object
 8   west          26280 non-null  object
 9   system        26280 non-null  object
dtypes: object(10)
memory usage: 2.0+ MB


In [91]:
# dropping commas for data consistency
load_df = load_df.replace({',' : ''}, regex=True)

In [93]:
# saving converted dataframe to a new csv file to be combined with the energy generation and temperature files
load_df.to_csv('/Users/tovi/Downloads/load_data_converted.csv', index=False)

In [ ]:
"""
CREATE TABLE load_data (timestamp TIMESTAMP NOT NULL,
						coast_load DECIMAL(10,2),
						east_load DECIMAL(10,2),
						farwest_load DECIMAL(10,2),
						north_load DECIMAL(10,2),
						northcentral_load DECIMAL(10,2),
						south_load DECIMAL(10,2),
						southcentral_load DECIMAL(10,2),
						west_load DECIMAL(10,2),
						system_load DECIMAL(10,2)
						);
"""

DECIMAL precision and scale was set to (10,2) to capture the larger values. Upload of the CSV resulted in 26,280 rows being returned, which is in line with the amount expected.

## Combining CSV Files Using SQL

In [ ]:
"""
SELECT wind_data.timestamp, temp_2m_south, wind_u_south, wind_v_south, temp_2m_north, wind_u_north, wind_v_north, temp_2m_east, wind_u_east,  wind_v_east, temp_2m_west, wind_u_west, wind_v_west, coast_load, east_load, farwest_load, north_load, northcentral_load, south_load, southcentral_load, west_load, system_load, wind_system, wind_coast, wind_south, wind_west, wind_north  
FROM wind_data
LEFT JOIN wind_temp_data
ON wind_data.timestamp = wind_temp_data.time
LEFT JOIN load_data
ON wind_data.timestamp = load_data.timestamp
ORDER BY wind_data.timestamp;
"""

This join returns all records in the wind_data table alongside the corresponding temperature and load values for matching timestamps from the temperature and load tables. Returns 25466 rows (check for duplicates in wind_data file).

In [98]:
wind_temp = pd.read_csv('/Users/tovi/Downloads/wind_temp.csv', quotechar="'")
wind_temp.head()

timestamp  temp_2m_south  wind_u_south  wind_v_south  \
0  2021-01-01 01:00:00          44.45      8.833785     -4.142875   
1  2021-01-01 02:00:00          43.45      8.720156     -3.938205   
2  2021-01-01 03:00:00          41.90      8.357309     -3.723078   
3  2021-01-01 04:00:00          41.48      8.048112     -3.574828   
4  2021-01-01 05:00:00          40.61      7.704257     -3.411989   

   temp_2m_north  wind_u_north  wind_v_north  temp_2m_east  wind_u_east  \
0          36.22     -0.381892     -7.454720         51.64     9.073054   
1          36.13      0.146601     -6.802784         49.33     8.665540   
2          35.77      0.887693     -6.319172         46.58     7.918151   
3          35.56      1.753668     -5.496035         46.07     8.195437   
4          35.70      2.519655     -4.641656         45.62     8.760590   

   wind_v_east  ...  northcentral_load  south_load  southcentral_load  \
0     5.087430  ...           13340.34     3586.55            7693.74   
1     4.598661  ...           13311.63     3586.12            7566.95   
2     4.829106  ...           13356.44     3570.59            7506.69   
3     5.092289  ...           13528.65     3590.90            7563.97   
4     4.992007  ...           13856.24     3647.15            7716.80   

   west_load  system_load  wind_system  wind_coast  wind_south  wind_west  \
0    1421.39     43321.05     14012.32     1975.35     1246.65    7623.01   
1    1405.54     43063.07     14315.26     2023.05      980.89    7778.31   
2    1402.28     43090.06     14247.88     1819.02      742.39    8068.91   
3    1405.09     43486.59     13273.42     1426.17      750.18    7543.59   
4    1411.72     44265.32     12248.01     1280.43      825.16    6667.16   

   wind_north  
0      146.97  
1      174.87  
2      211.95  
3      186.87  
4      169.53  

[5 rows x 27 columns]

In [100]:
# saving complete file
wind_temp.to_csv('/Users/tovi/Downloads/wind_data_combined.csv', index=False)

In [103]:
df = pd.read_csv('/Users/tovi/Downloads/wind_data_combined.csv')
df.head()

timestamp  temp_2m_south  wind_u_south  wind_v_south  \
0  2021-01-01 01:00:00          44.45      8.833785     -4.142875   
1  2021-01-01 02:00:00          43.45      8.720156     -3.938205   
2  2021-01-01 03:00:00          41.90      8.357309     -3.723078   
3  2021-01-01 04:00:00          41.48      8.048112     -3.574828   
4  2021-01-01 05:00:00          40.61      7.704257     -3.411989   

   temp_2m_north  wind_u_north  wind_v_north  temp_2m_east  wind_u_east  \
0          36.22     -0.381892     -7.454720         51.64     9.073054   
1          36.13      0.146601     -6.802784         49.33     8.665540   
2          35.77      0.887693     -6.319172         46.58     7.918151   
3          35.56      1.753668     -5.496035         46.07     8.195437   
4          35.70      2.519655     -4.641656         45.62     8.760590   

   wind_v_east  ...  northcentral_load  south_load  southcentral_load  \
0     5.087430  ...           13340.34     3586.55            7693.74   
1     4.598661  ...           13311.63     3586.12            7566.95   
2     4.829106  ...           13356.44     3570.59            7506.69   
3     5.092289  ...           13528.65     3590.90            7563.97   
4     4.992007  ...           13856.24     3647.15            7716.80   

   west_load  system_load  wind_system  wind_coast  wind_south  wind_west  \
0    1421.39     43321.05     14012.32     1975.35     1246.65    7623.01   
1    1405.54     43063.07     14315.26     2023.05      980.89    7778.31   
2    1402.28     43090.06     14247.88     1819.02      742.39    8068.91   
3    1405.09     43486.59     13273.42     1426.17      750.18    7543.59   
4    1411.72     44265.32     12248.01     1280.43      825.16    6667.16   

   wind_north  
0      146.97  
1      174.87  
2      211.95  
3      186.87  
4      169.53  

[5 rows x 27 columns]